flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java

```diff
@@ -0,0 +1,241 @@
+ * on the given lock object. This ensures that we don't call methods on a
+ * {@link OneInputStreamOperator} concurrently with the timer callback or other things.
+ *
+ * @param <IN> The type of the record that can be read with this record reader.
+ */
+@Internal
+public final class StreamOneInputProcessor<IN> implements StreamInputProcessor {
+
+	private static final Logger LOG = LoggerFactory.getLogger(StreamOneInputProcessor.class);
+
+	private final StreamTaskInput input;
+
+	private final Object lock;
+
+	private final OperatorChain<?, ?> operatorChain;
+
+	// ---------------- Status and Watermark Valve ------------------
+
+	/** Valve that controls how watermarks and stream statuses are forwarded. */
+	private StatusWatermarkValve statusWatermarkValve;
+
+	private final StreamStatusMaintainer streamStatusMaintainer;
+
+	private final OneInputStreamOperator<IN, ?> streamOperator;
+
+	// ---------------- Metrics ------------------
+
+	private final WatermarkGauge watermarkGauge;
+	private Counter numRecordsIn;
+
+	@SuppressWarnings("unchecked")
+	public StreamOneInputProcessor(
+			InputGate[] inputGates,
+			TypeSerializer<IN> inputSerializer,
+			StreamTask<?, ?> checkpointedTask,
+			CheckpointingMode checkpointMode,
+			Object lock,
+			IOManager ioManager,
+			Configuration taskManagerConfig,
+			StreamStatusMaintainer streamStatusMaintainer,
+			OneInputStreamOperator<IN, ?> streamOperator,
+			TaskIOMetricGroup metrics,
+			WatermarkGauge watermarkGauge,
+			String taskName,
+			OperatorChain<?, ?> operatorChain) throws IOException {
+
+		InputGate inputGate = InputGateUtil.createInputGate(inputGates);
+
+		CheckpointBarrierHandler barrierHandler = InputProcessorUtil.createCheckpointBarrierHandler(
+			checkpointedTask,
+			checkpointMode,
+			ioManager,
+			inputGate,
+			taskManagerConfig,
+			taskName);
+		this.input = new StreamTaskNetworkInput(barrierHandler, inputSerializer, ioManager, 0);
+
+		this.lock = checkNotNull(lock);
+
+		this.streamStatusMaintainer = checkNotNull(streamStatusMaintainer);
+		this.streamOperator = checkNotNull(streamOperator);
+
+		this.statusWatermarkValve = new StatusWatermarkValve(
+			inputGate.getNumberOfInputChannels(),
+			new ForwardingValveOutputHandler(streamOperator, lock));
+
+		this.watermarkGauge = watermarkGauge;
+		metrics.gauge("checkpointAlignmentTime", barrierHandler::getAlignmentDurationNanos);
+
+		this.operatorChain = checkNotNull(operatorChain);
+	}
+
+	@Override
+	public boolean isFinished() {
+		return input.isFinished();
+	}
+
+	@Override
+	public CompletableFuture<?> isAvailable() {
+		return input.isAvailable();
+	}
+
+	@Override
+	public boolean processInput() throws Exception {
+		initializeNumRecordsIn();
+
+		StreamElement recordOrMark = input.pollNextNullable();
+		if (recordOrMark != null) {
+			int channel = input.getLastChannel();
+			checkState(channel != StreamTaskInput.UNSPECIFIED);
+
+			processElement(recordOrMark, channel);
+		}
+		checkFinished();
```

Comment:
Any reason why we need to check this in both cases and not just `recordOrMark == null`? It uses `synchronized` so saving calls might be good.

Reply:
But it only does it if the input `isFinished()`, so that should be a non issue.